### PR TITLE
[wicket] command-line interface to mupdate

### PR DESCRIPTION
**Pull request by sunshowers**
## Overview

Implement a command-line interface to mupdates, with an eye to using this in CI.

To achieve this, add two commands to wicket:

### `rack-update start`

This command starts mupdates on one or more components. The help looks like:

```
Start one or more running updates

Usage: wicket rack-update start [OPTIONS]

Options:
  -h, --help  Print help (see more with '--help')

Component selectors:
      --sled <SLED>      The sleds to operate on
      --switch <SWITCH>  The switches to operate on
      --psc <PSC>        The PSCs to operate on

Update options:
      --force-update-rot  Force update the RoT even if the version is the same
      --force-update-sp   Force update the SP even if the version is the same
  -d, --detach            Detach after starting the update

Global options:
      --color <COLOR>  Color output [default: auto] [possible values: auto, always, never]
```

### `rack-update attach`

This command attaches to any existing mupdates on one or more components. The help looks like:

```
Attach to one or more running updates

Usage: wicket rack-update attach [OPTIONS]

Options:
  -h, --help  Print help (see more with '--help')

Component selectors:
      --sled <SLED>      The sleds to operate on
      --switch <SWITCH>  The switches to operate on
      --psc <PSC>        The PSCs to operate on

Global options:
      --color <COLOR>  Color output [default: auto] [possible values: auto, always, never]
```

### What does this look like?

Here's a screenshot:

image

Buildomat doesn't yet support ANSI escape codes (https://github.com/oxidecomputer/buildomat/issues/40), but if we add support for them then we'd get full color support. Otherwise, it'll look like this without colors.

## Implementation

This is a somewhat large PR but almost all added code. The bulk of added code is in the update-engine, which has two new displayers:

### 1. Line display

This displayer shows events line by line. Since this is meant to be used in CI, we have a hard constraint that we can't go back and change old lines.

### 2. Group display

This displayer shows events across several possible executions. This is just a slightly more involved version of the line display, and shares most of its code (via `line_display_shared.rs`).

The rest of this PR:

* Hooks up a group displayer to wicket, which is pretty straightforward.
* Adds some more to the `update-engine-basic` example.

## Testing this

The wicket side can be tested with:

```
SSH_ORIGINAL_COMMAND='rack-update start --sled 1' cargo run -p wicket
```

For a more comprehensive example, run `cargo run --example update-engine-basic -- --display-style group`.

Depends on #4429 (and a number of other PRs that have already landed.)

## Future work

One bit of work that hasn't been done yet is doing an integration test with wicketd, wicket, MGS and sp-sim. One issue is that currently, we don't generate machine-readable output from wicket (maybe we should!), and given that that's not the case we only know how to exit with either the exit code 0 (success) or 1 (failure).

### CODE REVIEW

**Comment by benjaminleonard** [comment]
Let me know if you'd like to collab on the look of it 👀 
Will it run inside of wicket?

**Comment by sunshowers** [comment]
> Let me know if you'd like to collab on the look of it 👀

This would be cool! Let's chat soon.

> Will it run inside of wicket?

It won't be part of the TUI but it will be invoked via the wicket binary.

File: update-engine/src/display/group_display.rs [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+            Err(UnknownReportKey {})
+        }
+    }
+
+    /// Writes a "Status" or "Summary" line to the writer with statistics.
+    pub fn write_stats(&mut self, header: &str) -> std::io::Result<()> {
+        // Add a prefix which is equal to the maximum width of the prefixes.
+        // [prefix 00:00:00] takes up self.max_width + 9 characters inside the
+        // brackets.
+        let prefix = " ".repeat(self.max_width);
```
Comment (jgallagher) [inline update-engine/src/display/group_display.rs:154]: Should this be `self.max_width + 9`? If not, I don't follow the comment. :)

File: wicket/src/cli/rack_update.rs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            }
+            Err(error) => {
+                // Error responses can be printed out more clearly.
+                if let wicketd_client::Error::ErrorResponse(rv) = &error {
+                    slog::error!(
+                        log,
+                        "Error response from wicketd: {}",
+                        rv.message
+                    );
+                    bail!("Received error from wicketd while starting update");
```
Comment (jgallagher) [inline wicket/src/cli/rack_update.rs:122]: Should we include `rv.message` in this `bail!`? Or is `log` visible to the caller and not logging to `/tmp` or wherever the default wicket log ends up?

**Review by jgallagher — APPROVED** [review]
This looks really slick :+1: Much nicer than the hokey line display I used for `preflight uplink` (and now I want to try using this for preflight!).

File: update-engine/examples/update-engine-basic/main.rs [hunk 3]
```diff
@@ -146,9 +200,30 @@ impl ExampleContext {
...
+                    // Now indicate a progress reset.
+                    cx.send_progress(StepProgress::reset(
+                        serde_json::Value::Null,
+                        "Progress reset",
+                    ))
+                    .await;
+
+                    // Try again.
+                    let mut buf_list = BufList::new();
+                    for i in 0..8 {
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply update-engine/examples/update-engine-basic/main.rs:226]: Yeah, fixed. Thanks!

File: update-engine/src/buffer.rs [hunk 4]
```diff
@@ -249,13 +266,37 @@ impl<S: StepSpec> EventStore<S> {
...
+                            Some((parent_key, child_index))
+                        }
+                        None => {
+                            // This should never happen -- it indicates that the
+                            // parent key was unknown. This can happen if we
+                            // didn't receive an event regarding a parent
+                            // execution being started.
+                            //
+                            // TODO: This should probably be an error that gets
+                            // bubbled up to callers.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply update-engine/src/buffer.rs:293]: No, we can't really do anything useful here. I'll just remove this comment.

File: update-engine/src/display/line_display_shared.rs [hunk 5]
```diff
@@ -0,0 +1,1017 @@
...
+    ) {
+        match &step_event.kind {
+            StepEventKind::NoStepsDefined => {
+                let mut line = self
+                    .formatter
+                    .start_line(self.prefix, Some(step_event.total_elapsed));
+                swrite!(
+                    line,
+                    "{:>HEADER_WIDTH$} ",
+                    "No steps defined"
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply update-engine/src/display/line_display_shared.rs:142]: I just removed the `{:>HEADER_WIDTH$} ` condition, thanks for catching that.

File: update-engine/src/display/line_display_shared.rs [hunk 6]
```diff
@@ -0,0 +1,1017 @@
...
+            }
+            StepEventKind::Nested { step, event, .. } => {
+                // Look up the child event's ID to add to the nest data.
+                let child_step_key = StepKey {
+                    execution_id: event.execution_id,
+                    // TODO: we currently look up index 0 because that should
+                    // always exist (unless no steps are defined, in which case
+                    // we skip this). The child index is actually shared by all
+                    // steps within an execution. Fix this by changing
+                    // EventBuffer to also track general per-execution data.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply update-engine/src/display/line_display_shared.rs:418]: No, probably will do this as part of tech debt at some point. It's not too bad at the moment.

File: update-engine/src/display/group_display.rs (hunk 1 above)
Comment (sunshowers) [reply update-engine/src/display/group_display.rs:154]: The code is correct but the comment is confusing. Will rewrite the comment.

File: wicket/src/cli/rack_update.rs (hunk 2 above)
Comment (sunshowers) [reply wicket/src/cli/rack_update.rs:122]: For interactive use, things are logged to both stderr and the wicket log file -- so this is fine.

**Pull request merged**